### PR TITLE
tlauncher: init at 2.839

### PR DESCRIPTION
**Pull request by lucasew**
###### Motivation for this change
- Missing package
- Solves #114571 
- Dealing with quirks caused by how TLauncher and Minecraft bootstraps it's JVMs (Shiginima hasn't this problem)

###### Things done
- Packaged the tlauncher Minecraft launcher 

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by thiagokokada** [comment]
Result of `nixpkgs-review pr 152744` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>tlauncher</li>
  </ul>
</details>

**Review by thiagokokada — APPROVED** [review]
Code-wise LGTM, and the launcher works fine.

Would like another review from someone who actually knows about Minecraft, thought.

File: pkgs/top-level/all-packages.nix [hunk 1]
```diff
@@ -32079,6 +32079,8 @@ with pkgs;
   };
   tlaplusToolbox = callPackage ../applications/science/logic/tlaplus/toolbox.nix {};
 
+  tlauncher = callPackage ../games/tlauncher/default.nix {};
```
Comment (SuperSandro2000) [inline pkgs/top-level/all-packages.nix:32082]:
```suggestion
  tlauncher = callPackage ../games/tlauncher { };
```

File: pkgs/games/tlauncher/default.nix [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+in stdenvNoCC.mkDerivation {
+  pname = "tlauncher";
+  inherit version;
+
+  dontUnpack = true;
+
+  installPhase = ''
+    mkdir $out/{bin,share/applications} -p
+    cp ${fhs}/bin/tlauncher $out/bin
+    cp ${desktopItem}/share/applications/* $out/share/applications
```
Comment (SuperSandro2000) [inline pkgs/games/tlauncher/default.nix:79]: Please use copyDesktopItems.

File: pkgs/games/tlauncher/default.nix [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+  desktopItem = makeDesktopItem {
+    name = "tlauncher";
+    exec = "tlauncher";
+    # icon = "tlauncher"; # TODO: add icon
+    comment = "Minecraft launcher";
+    desktopName = "TLauncher";
+    categories = "Game;";
+  };
+in stdenvNoCC.mkDerivation {
+  pname = "tlauncher";
```
Comment (SuperSandro2000) [inline pkgs/games/tlauncher/default.nix:71]:
```suggestion
  pname = "tlauncher-wrapper";
```

File: pkgs/games/tlauncher/default.nix [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+, openjdk8
+, buildFHSUserEnv
+, fetchzip
+, copyDesktopItems
+, makeDesktopItem
+}:
+let
+  version = "2.839";
+  src = stdenvNoCC.mkDerivation {
+    name = "tlauncher-${version}.jar";
```
Comment (SuperSandro2000) [inline pkgs/games/tlauncher/default.nix:12]:
```suggestion
    pname = "tlauncher";
    inherit version;
```

File: pkgs/games/tlauncher/default.nix (hunk 2 above)
Comment (lucasew) [reply pkgs/games/tlauncher/default.nix]: for some reason it's not working

i will do a more deep look later

happy new year

**Comment by thiagokokada** [comment]
Result of `nixpkgs-review pr 152744` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>tlauncher</li>
  </ul>
</details>

**Review by thiagokokada — APPROVED** [review]
Not sure if this actually enters in-game since I don't play Minecraft, however at least the launcher works fine.

**Comment by lucasew** [comment]
In my machine it works awesomely well

I had played with the boys on a modded survival server with optifine and a non stock shader and it worked very well.

**Comment by Fr3shTea** [comment]
Isn't TLauncher is a Minecraft launcher that allows to play cracked Minecraft? Is this kind of software accepted into nixpkgs?

**Comment by bryanasdev000** [comment]
> Isn't TLauncher is a Minecraft launcher that allows to play cracked Minecraft (without actually buying the game)? Is this kind of software is allowed in nixpkgs?

Well, you can play the game that you bought too (java version), so I think it's the same "gray zone" as emulator for example.

**Comment by xdelsy** [comment]
What's worse, they have been seen deleting competitors' servers from servers.dat - the local database of the user's saved servers in Minecraft - and adding other servers that paid them, without the user's knowledge. There's [an article](https://minecraftmain.ru/2018/02/tlauncher-story) about this from Igor Fromgate, the proofreader of the Russian translation of Minecraft. Use [DeepL](https://deepl.com) for a better translation.

We are particularly interested in the two parts of the article entitled "Server Ads in TLauncher" (`Реклама серверов в TLauncher`)  and "Blacklisting of Servers" (`Чёрный список серверов`):

> Advertising sales are handled by "manager Alice" and here are the current advertising rates _(note: the following is a partial translation of the administration messages from 2018; the conversion of roubles into dollars is done by me)_:
> 
>> There is an opportunity to place in the second tab on the day.
>> 
>> - one weekday: 1000 rubles _(13.10 USD)_
>> - one weekend day: 1200 rubles _(15.71 USD)_
>> - a week: 5000 rubles _(65.48 USD)_
>> 
>> On the main (in the tab "Recommended servers"):
>> 
>> - one weekday: 2500 rubles _(32.74 USD)_
>> - one weekend day: 3000 rubles _(39.29 USD)_

> In addition to the "add server to server list" feature, TLauncher also hides the opposite feature. The launcher can remove certain addresses from the server list. This blacklist is loaded by the launcher from the following links _(note: the files are still present)_:
> 
> - http://repo.tlauncher.org/update/downloads/configs/servers.json
> - https://tlauncher.org/repo/update/downloads/configs/servers.json
> - http://ru-m.org/update/downloads/configs/servers.json
>
> At the time of writing, there were 13 servers on these lists _(note: including duplicates, 135 as of today)_.
>
> If you add any of the servers to Minecraft and then restart it, TLauncher will neatly remove those servers from the list the next time you start the game. In other words, the next time the player starts the game, he will not find the server and will have to add it again or play on some other server.
>
> A source told me that the blacklist is used as a form of tying servers to advertisements. That is, when a server stops buying advertising, it is removed from the clients of the players who were playing on it in order to force the server owners to pay for advertising again. I couldn't find any confirmation of this theory, but what I found out from some of the server owners wasn't much better.

The article also tells of DDoS attacks, partnerships, and a few stories of former partners whose servers are now blacklisted.

That was a long time ago, four years ago. Their approach to business may have changed in that time, but I think it needs investigation. Sorry in advance, can't help with that at the moment, very busy lately.

**Comment by lucasew** [comment]
In my case, until they don't block aternos this doesn't seem to be a problem

BTW I did a script to ping check how many of those servers are online

```bash
#!/usr/bin/env nix-shell
#! nix-shell -i bash -p curl jq

curl "https://repo.tlauncher.org/update/downloads/configs/servers.json" 2> /dev/null | grep 'address' | sed 's/^[^:]*: "\([^":]*\)[^$]*$/\1/g' | tr '[:upper:]' '[:lower:]' | sort | uniq | while read -r line; do
    ping "$line" -c 1 > /dev/null 2> /dev/null && printf "OK" || printf "FAIL"
    printf " $line"
    printf "\n"
done
```

**Pull request merged**